### PR TITLE
feat(catalog): Phase B Stack over Phase A on Native Capture

### PR DESCRIPTION
**Pull request by zaoxing**
Stacked on #127 (Phase A). **Phase B complete** — the cold write path ported to native: B1, B2, B3, B4.

## Slices (one commit each)

- **B1a lease coordinator** — claim / head / renew / release / reject_if_gone / fence; SQL ported textually (client-side `%(name)s` substitution matching clickhouse-driver); byte-identity gate vs the Python module.
- **B1b version allocator** — sole-claimant `allocate_version`; contention test: racing processes → distinct versions, each durably claimed.
- **B2 descriptors + fenced publish** — `write_descriptors`, replay guard, `commit_packs`, `publish_snapshot` (chunked fenced manifest, barrier+fence+visibility as one statement, ownership read-back, race/conflict taxonomy, outcome-unknown quarantine). Taken-over publisher writes **nothing**; the statement-gap leaves exactly the documented inert orphans; head-to-head `capture_raw` rows byte-identical (MATERIALIZED facets included).
- **B3 indexer-native** — the `CatalogIndexer` loop with PackIndex.from_store validation ported (footer refused at the boundary as untrusted data); race retry rewrites descriptors at the winning version. E2E: native sink → uploader → index → **Python CaptureReader resolves every capture**; replay guard skips pass 2.
- **B4 schema + GC** — ensure/drop with the install serialised on the publisher lease (borrow-or-take, one-TTL budget, second initialiser completes by waiting), server-side conditional stamp, compatibility refusals (kinds, sort keys, the version-argument property, stamp version, inventory-without-membership, legacy-object-beside-this-build, CHECK GRANT visibility before any verdict); `collect_garbage` with the two-read settle floored at one publish timeout, literal-pair deletions, head/claims bounds, idempotent second pass.
- **Review pass** — config validation ported from `ClickHouseCatalogConfig` (identifiers, fence margin, quorum rules, whole-second publish timeout), four silent-loss bugs in the indexer/writer fixed with reproductions, schema refusal coverage added, warning surface to zero.

## Verification

`tests/test_native_catalog_lease_live.py` — **44 live tests**, manual+clickhouse, in the clickhouse-live job beside the Python suites. 44/44 across two consecutive runs. Whole live selector as CI runs it (`pytest tests/*_live.py -m "clickhouse and manual and not garage"`): **107 passed, 4 deselected, 0 skipped**. CPU gate **1202 passed**. Clean native rebuild carries **zero warnings** under `-Wall -Wextra`.

**`verify_replicated_quorum.py` vs the C++ writer: 12/12 PASS, both legs, two runs.** This was previously recorded here as "not runnable in this environment" — it was runnable. The harness keeper was already live, so only `tests/tools/quorum_harness/server.xml` needed starting (tcp 9010 / http 8110), and the verifier auto-runs its native leg when `native/build/conformance_catalog` exists. The six checks through the C++ writer match the six through Python: publish cycle with `insert_quorum=2`, settings recorded in the query log on every deciding INSERT, retention on ReplicatedMergeTree, the quorum-unset control, quorum-off-mid-life failing loudly, and an unsatisfiable quorum refused promptly with Code 285.

## Checkpoint B: PASS with two recorded gaps

Reviewed against this PR + #127, every criterion re-verified at the current tip rather than read off earlier commits. Recorded in `tasks/todo.md`.

Gaps, neither blocking, both written down rather than implied:

- **#125 client-side semantics are half-ported.** The quarantine half is real and tested; per-writer publish serialisation and process binding are not, and their three concurrency tests are absent. The native writer relies on the driver being one process running one op at a time. Sound for the shipped path (pybind and the driver are single-threaded per writer) and a real limit for any other embedding — `catalog_writer.h` now states it at the class it applies to.
- **The byte-identity gate covers `release` and `fence` only.** The parameterized statements (`committed_pack_ids` `IN`, the manifest `arrayJoin` INSERT) are semantically equivalent rather than textually identical — native wraps the rendered list in its own parens and omits the driver's comma-space. The risk table in `tasks/plan.md` now says which is which.

Still open and named at the checkpoint: the quiet-host re-measure, which is the only thing between the recorded throughput numbers and a published one. (The "reference host" is this same 5955WX, so the obligation is scheduling, not hardware. Quantified on 2026-09-06: three consecutive runs of one bench binary at load 14.5 gave 0.259 / 0.492 / 0.519 GiB/s -- the decision survives the worst reading, the figures are not publishable. Protocol in docs/benchmarks.md.)

## CI

The clickhouse-live job had been red since the B3 tests landed, on two counts — the second hidden behind the first. `botocore` was missing (the suite imports the fake-S3 fixture at module level, so its absence errors collection of all 44 tests rather than skipping the S3-backed ones), and the job built only `conformance_catalog` while five of those tests also drive `conformance_sink` and `conformance_store`. Both fixed; verified with the job's own command and against the junit XML its "a skip is an error" gate parses.

### CODE REVIEW

**Comment by zaoxing** [comment]
**Review pass landed (09e691a) + quorum verifier extended — Phase B fully verified.**

Three bugs found and fixed with regression tests:
1. Indexer erased failed packs mid-walk — packs behind a failure were published/committed as indexed without ever being read (committed-and-invisible, unrepairable).
2. The batch-byte guard sat inside the read's try — oversized batches got blamed on a pack and swallowed into a "successful" partial index.
3. A lapsed quarantine was never cleared — the writer refused forever and reported itself quarantined after the window.

Plus: `verify_replicated_quorum.py` now runs **both legs** — Python writer and native writer — against the Keeper-backed replicated harness: 12/12 checks PASS (verdict 0) twice. The native leg proves on ReplicatedMergeTree: the quorum-durable publish cycle, `insert_quorum=2 / parallel=0 / bounded timeout` recorded per the query log on every native deciding INSERT, retention mutations admitted, quorum-off mid-life failing loudly, and the unsatisfiable quorum refused with Code 285 promptly.

State: 27/27 live tests × 3 runs, CPU gate 1202 passed. Still awaiting human review here and on #127.

File: native/csrc/catalog/clickhouse_client.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        }
+        rendered.push_back(c);
+      }
+      rendered.push_back('\'');
+    }
+    const std::string needle = "%(" + name + ")s";
+    size_t at;
+    while ((at = query.find(needle)) != std::string::npos) {
+      query.replace(at, needle.size(), rendered);
+    }
```
[the remark this answers is not in the document]
Comment (zaoxing) [reply native/csrc/catalog/clickhouse_client.cpp:57]: Reproduced with a test first (test_a_placeholder_in_a_string_param_does_not_hang_the_substitution — a holder that literally says "%(holder)s" is a legal 11-byte string; the driver hung, and the test arms a timeout on the pipe read so it fails loudly instead of hanging the suite) and fixed in e6fd6b0: the search advances past each replacement, so a self-referential value is never re-matched inside the inserted text.

File: native/csrc/catalog/conformance_catalog.cpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        if (!first_row) out += ",";
+        first_row = false;
+        out += "[";
+        bool first_field = true;
+        for (const auto& field : row) {
+          if (!first_field) out += ",";
+          first_field = false;
+          out += "\"";
+          escape_into(field, &out);
+          out += "\"";
```
[the remark this answers is not in the document]
Comment (zaoxing) [reply native/csrc/catalog/conformance_catalog.cpp]: Already fixed by the parallel review pass on this file — rows_to_json now wraps fields with escape_into alone (EscapeJson emits its own quotes). Regression test added: test_execute_returns_valid_json_for_non_empty_results drives a SELECT with two fields and asserts the parsed result. Verified passing.

File: native/csrc/catalog/clickhouse_client.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  }
+  return query;
+}
+
+std::string url_encode(const std::string& value) {
+  char* escaped = curl_easy_escape(nullptr, value.c_str(),
+                                   static_cast<int>(value.size()));
+  const std::string out(escaped);
+  curl_free(escaped);
+  return out;
```
[the remark this answers is not in the document]
Comment (zaoxing) [reply native/csrc/catalog/clickhouse_client.cpp:72]: Fixed in e6fd6b0: curl_easy_escape returning null (allocation failure) now throws ClickHouseError instead of constructing a std::string from null. Not testable cheaply (OOM path) — fixed as hardening per the review.

File: native/csrc/catalog/lease_coordinator.cpp [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+bool LeaseCoordinator::fence_eval(const std::string& lease_id,
+                                  uint64_t publish_timeout_ns,
+                                  uint64_t clock_skew_ns) const {
+  const std::vector<Row> rows = client_->execute(
+      "SELECT " + fence(),
+      {{"lease_id", lease_id},
+       {"publish_timeout_ns", publish_timeout_ns},
+       {"clock_skew_ns", clock_skew_ns}});
```
[the remark this answers is not in the document]
Comment (zaoxing) [reply native/csrc/catalog/lease_coordinator.cpp]: Fixed in e6fd6b0: fence_eval now carries select_sequential_consistency — its answer decides whether the fence admits, and a replica behind on the lease table would admit or deny on stale head state. Same doctrine as DECIDING_READ in clickhouse_sql.py.

File: native/csrc/catalog/pack_index.cpp [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    for (const char c : dim) {
+      if (c == '-' && !any) {
+        negative = true;
+        continue;
+      }
+      if (c < '0' || c > '9') format_error("invalid shape dimension");
+      any = true;
+      value = value * 10 + static_cast<uint64_t>(c - '0');
+    }
+    if (!any) format_error("invalid shape dimension");
```
Comment (Samfisheryu) [inline native/csrc/catalog/pack_index.cpp:101]: [P2] Treat an empty shape array as a scalar, not an invalid dimension

For the legal scalar shape `[]`, `SplitElements(Unwrap(shape_json))` produces an empty token. This loop throws `invalid shape dimension`, instead of returning the scalar product of 1.

Reproduced through native sink → uploader → native indexer using `dtype='float32'`, `shape=()`, and a four-byte float payload. The pack writes and uploads successfully; Python `PackReader` reads the scalar descriptor correctly. Native indexing returns `ok=true` but `result.failed_packs=1`, `indexed_packs=0`, and the capture never appears in search.

Please handle `[]` as zero dimensions with product 1. Add a scalar pack to the live indexer test and assert per-pack failures are empty, not only the top-level `ok` flag.

File: native/csrc/catalog/clickhouse_client.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    // The search advances PAST each replacement: a rendered value that
+    // itself contains the placeholder (a holder that literally says
+    // "%(holder)s" is a legal 11-byte string) must not be re-matched
+    // inside the inserted text — restarting at the top grew without
+    // bound and hung the driver.
+    size_t from = 0;
+    size_t at;
+    while ((at = query.find(needle, from)) != std::string::npos) {
+      query.replace(at, needle.size(), rendered);
+      from = at + rendered.size();
```
Comment (Samfisheryu) [inline native/csrc/catalog/clickhouse_client.cpp:56]: [P2] Do not substitute another parameter inside an already-inserted value

Advancing `from` fixes the previously reported self-reference hang, but the next iteration of the outer parameter loop starts at zero on the modified query. It can still replace placeholders inside a value inserted by an earlier parameter.

Live ClickHouse repro using the existing catalog driver:

```python
driver.call(op='acquire', holder='worker-%(ttl_ns)s')
# Query the holder column in <prefix>_publisher_lease.
# Expected: 'worker-%(ttl_ns)s'
# Actual:   'worker-30000000000'
```

The literal holder string is legal; its content should not be interpreted as SQL template syntax. Please render placeholders by scanning the original template once, so parameter values are never revisited. This is a remaining case after the self-placeholder fix, not the same resolved hang.

**Review by Samfisheryu — COMMENTED** [review]
Two additional reproducible correctness issues are inline: legal scalar packs fail indexing, and SQL parameter values can still be rewritten by a different parameter's substitution. These were tested in the #129 stack at a4d3dcb and traced to this PR's unchanged implementation. The already-resolved self-placeholder hang is not being reported again; the second finding is a remaining cross-placeholder case.
